File: legacy/src/scenes/dungeon/economy/ground-items.ts
/**
 * Ground-item substrate — the one funnel every item on the floor is removed
 * through, plus the runtime drop-id sequence.
 *
 * Extracted verbatim from core.ts. This is deliberately the LOWEST layer of the
 * economy: coins, loot drops and pickups all remove items through
 * `removeGroundItem`, so it cannot live in any one of them.
 */
import { state } from "../state";
import { clearPile } from "../corpse-run";
import { coopItemTaken } from "../coop";

/** Runtime-drop network-id sequence (cards/potions/materials the authority
 * rolls mid-floor). Reset per floor beside zombieNidSeq. */
let itemNidSeq = 0;

/** Next runtime drop id. Bumped per drop; `resetItemNid()` zeroes it per floor. */
export function nextItemNid(): string {
  return "d" + itemNidSeq++;
}

/** Per-floor reset, called beside zombieNidSeq. */
export function resetItemNid(): void {
  itemNidSeq = 0;
}

/** Pull a ground item out of the world: unparent, free its GPU resources, drop
 * it from the list. Everything that removes an item goes through this — which
 * makes it the one funnel for co-op TAKE broadcasts: picking up a shared (nid'd)
 * item tells the floor so it vanishes on every screen. */
export function removeGroundItem(k: number): void {
  const it = state.groundItems[k];
  if (!it) return;
  coopItemTaken(it); // no-op for coins/personal drops (no nid) or offline
  state.scene?.remove(it.sprite.mesh);
  it.sprite.dispose();
  state.groundItems.splice(k, 1);
  // A corpse pile is only DONE when its last item is off the floor. Clearing it
  // on the first pickup would strand the rest on a refresh; clearing it here
  // means an interrupted recovery (you grabbed the sword, then died again)
  // leaves the remainder recoverable, which is the whole promise.
  if (it.corpseId && !state.groundItems.some((g) => g.corpseId === it.corpseId)) {
    clearPile(it.corpseId);
  }
}

File: Rust/crates/pk-core/src/economy/ground_items.rs
//! Ground-Item Substrate — Unified removal funnel, drop network id sequence, and corpse pile lifecycle.
//!
//! PORTS: `economy/ground-items.ts`

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct GroundItemRecord {
    pub id: u32,
    pub nid: Option<String>,
    pub corpse_id: Option<u32>,
}

#[derive(Clone, Debug, PartialEq, Eq, Default)]
pub struct GroundItemRegistry {
    pub items: Vec<GroundItemRecord>,
    pub item_nid_seq: u32,
}

impl GroundItemRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    /// Next runtime drop id. Bumped per mid-floor authority roll (e.g. "d0", "d1").
    pub fn next_item_nid(&mut self) -> String {
        let nid = format!("d{}", self.item_nid_seq);
        self.item_nid_seq += 1;
        nid
    }

    /// Zeroes the runtime drop sequence counter per floor.
    pub fn reset_item_nid(&mut self) {
        self.item_nid_seq = 0;
    }

    /// Adds a ground item to the registry.
    pub fn add_item(&mut self, item: GroundItemRecord) {
        self.items.push(item);
    }

    /// Pulls a ground item out of the registry and returns it.
    pub fn remove_ground_item(&mut self, index: usize) -> Option<GroundItemRecord> {
        if index < self.items.len() {
            Some(self.items.remove(index))
        } else {
            None
        }
    }

    /// Returns true if all ground items associated with a corpse pile have been removed.
    pub fn is_corpse_pile_cleared(&self, corpse_id: u32) -> bool {
        !self.items.iter().any(|it| it.corpse_id == Some(corpse_id))
    }
}
